File: c/problem_29.c
#include "problem_29.h"
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <assert.h>

void rleEncode(const char *in, char *out) {
    int n = (int)strlen(in);
    if (n == 0) { out[0] = '\0'; return; }
    int pos = 0, i = 0;
    while (i < n) {
        int count = 1;
        while (i + count < n && in[i + count] == in[i]) count++;
        pos += sprintf(out + pos, "%d%c", count, in[i]);
        i += count;
    }
    out[pos] = '\0';
}

void rleDecode(const char *in, char *out) {
    int pos = 0;
    const char *p = in;
    while (*p) {
        char *end;
        int count = (int)strtol(p, &end, 10);
        memset(out + pos, *end, count);
        pos += count;
        p = end + 1;
    }
    out[pos] = '\0';
}

void problem29(void) {
    printf("Problem 29:\n");
    char buf[256];
    rleEncode("AAAABBBCCDAA", buf);
    assert(strcmp(buf, "4A3B2C1D2A") == 0);
    rleDecode("4A3B2C1D2A", buf);
    assert(strcmp(buf, "AAAABBBCCDAA") == 0);
    rleEncode("A", buf);
    assert(strcmp(buf, "1A") == 0);
    rleDecode("1A", buf);
    assert(strcmp(buf, "A") == 0);
    rleEncode("", buf);
    assert(strcmp(buf, "") == 0);
    printf("All Tests Passed\n");
}

File: rust/problem_29.rs
pub fn rle_encode(s: &str) -> String {
    if s.is_empty() { return String::new(); }
    let chars: Vec<char> = s.chars().collect();
    let mut result = String::new();
    let mut count = 1usize;
    for i in 1..chars.len() {
        if chars[i] == chars[i - 1] {
            count += 1;
        } else {
            result.push_str(&count.to_string());
            result.push(chars[i - 1]);
            count = 1;
        }
    }
    result.push_str(&count.to_string());
    result.push(*chars.last().unwrap());
    result
}

pub fn rle_decode(s: &str) -> String {
    let chars: Vec<char> = s.chars().collect();
    let mut result = String::new();
    let mut i = 0;
    while i < chars.len() {
        let mut j = i;
        while j < chars.len() && chars[j].is_ascii_digit() { j += 1; }
        let count: usize = chars[i..j].iter().collect::<String>().parse().unwrap();
        let ch = chars[j];
        for _ in 0..count { result.push(ch); }
        i = j + 1;
    }
    result
}

pub fn problem29() {
    println!("Problem 29:");
    assert_eq!(rle_encode("AAAABBBCCDAA"), "4A3B2C1D2A");
    assert_eq!(rle_decode("4A3B2C1D2A"),  "AAAABBBCCDAA");
    assert_eq!(rle_encode("A"),           "1A");
    assert_eq!(rle_decode("1A"),          "A");
    assert_eq!(rle_encode(""),            "");
    println!("All Tests Passed");
}
